### DISCUSSION TITLE
results of trying Avalonia (and others) on GhostBSD (FreeBSD)

### DISCUSSION THREAD

**Discussion by sjsepan3** [AvaloniaUI/Avalonia#15853] [General]
## Overview

Although FreeBSD it is not supported, since I am able to run a port of Code-OSS and also of .Net SDK 8.0 on GhostBSD 2024040.1 (FreeBSD 14.0-STABLE), I thought I'd try porting some of my .Net projects from Linux (Mint).

- A simple **console** app builds/runs OK
- An **Avalonia** app gets build warnings about obsolete (deprecated) dialog calls (as it does on Linux) but no errors (the Linux app runs OK) -- but running  on GhostBSD gives "No runtime platform services configured." I've looked at  "Cannot start Avalonia under FreeBSD #6990 ", and I am not referencing Avalonia.Native (but should I?)  **Update**: release of 11.1.0, no change (didn't expect one, just following up)
- A **Terminal.Gui** app builds OK, but gets runtime error "Curses failed to initialize, the exception is: Error loading native library "libncursesw.so.6, libncursesw.so.5", even after ncurses library 6.4_1 installed. **Update**:  If so, put 
 ```
 Application.UseSystemConsole = true;
 ```
  before 
 ```
  Application.Init();
```
- A **Gtk.Sharp** app build has no warnings or errors, but at runtime: "The type initializer for 'Gtk.Application' threw an exception." even after installing the gtk-sharp30, gtk-sharp2.0, gio-sharp pkgs
- A **Modern.Forms** app build has no warnings or errors, but at runtime: "The type initializer for 'Modern.Forms.Form' threw an exception."

I'm not really expecting a fix; just putting this here for future reference/discussion. The Avalonia Linux app I used is this one [https://gitlab.com/sjsepan/AvaloniaApp1](https://gitlab.com/sjsepan/AvaloniaApp1)

## Code-OSS

```
Version: 1.88.1
Commit: e170252f762678dec6ca2cc69aba1570769a5d39
Date: 2024-05-07T13:49:21.988Z
Electron: 28.3.1
ElectronBuildId: undefined
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: FreeBSD x64 14.0-STABLE
```

## dotnet --info

```
.NET SDK:
 Version:           8.0.105
 Commit:            eae90abaaf
 Workload version:  8.0.100-manifests.796a77f8

Runtime Environment:
 OS Name:     FreeBSD
 OS Version:  14
 OS Platform: FreeBSD
 RID:         freebsd-x64
 Base Path:   /usr/home/sjsepan/.dotnet/sdk/8.0.105/

.NET workloads installed:
 Workload version: 8.0.100-manifests.796a77f8
There are no installed workloads to display.

Host:
  Version:      8.0.5
  Architecture: x64
  Commit:       087e15321b

.NET SDKs installed:
  6.0.403 [/usr/home/sjsepan/.dotnet/sdk]
  8.0.105 [/usr/home/sjsepan/.dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.11 [/usr/home/sjsepan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.5 [/usr/home/sjsepan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.11 [/usr/home/sjsepan/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.5 [/usr/home/sjsepan/.dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/home/sjsepan/.dotnet]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

**Comment by kekekeks** [comment]
Avalonia.X11 currently depends on `epoll` for its event loop. Might need an alternative implementation fro BSD-based platforms

**Reply by kekekeks** [reply]
There are also plans to use timerfd_* apis, not sure if those are available with BSD

**Reply by sjsepan3** [reply]
> Avalonia.X11 currently depends on `epoll` for its event loop. Might need an alternative implementation fro BSD-based platforms

Ah, I just ran across an older issue where you already discussed this:
https://github.com/AvaloniaUI/Avalonia/issues/6990

I thought I'd searched for BSD-related issues, but I missed that one.

**Reply by ok-coder1** [reply]
> Avalonia.X11 currently depends on `epoll` for its event loop. Might need an alternative implementation fro BSD-based platforms

What about trying to use [FreeBSDDesktop/epoll-shim](https://github.com/FreeBSDDesktop/epoll-shim)?
